File: apple/Sources/CCDAEngine/Models/Media/CCDAMediaType.swift
import Foundation

/// Known media types with fallback for vendor-specific MIME values.
public enum CCDAMediaType: Hashable, Sendable {
    case imagePNG
    case imageJPEG
    case imageGIF
    case applicationPDF
    case textPlain
    case textHTML
    case unsupported(String)
    case unknown

    /// Creates a media type from a MIME string.
    public init(mimeType: String?) {
        guard let mimeType, !mimeType.isEmpty else {
            self = .unknown
            return
        }

        switch mimeType.lowercased() {
        case "image/png":
            self = .imagePNG
        case "image/jpeg", "image/jpg":
            self = .imageJPEG
        case "image/gif":
            self = .imageGIF
        case "application/pdf":
            self = .applicationPDF
        case "text/plain":
            self = .textPlain
        case "text/html":
            self = .textHTML
        default:
            self = .unsupported(mimeType)
        }
    }

    /// MIME string for this media type when known.
    public var mimeType: String? {
        switch self {
        case .imagePNG:
            return "image/png"
        case .imageJPEG:
            return "image/jpeg"
        case .imageGIF:
            return "image/gif"
        case .applicationPDF:
            return "application/pdf"
        case .textPlain:
            return "text/plain"
        case .textHTML:
            return "text/html"
        case .unsupported(let value):
            return value
        case .unknown:
            return nil
        }
    }

    /// Whether this type is an image MIME type.
    public var isImage: Bool {
        mimeType?.hasPrefix("image/") == true
    }
}

File: android/ccda-engine/src/main/kotlin/com/ccdakit/engine/Models/Media/CCDAMediaType.kt
package com.ccdakit.engine

/** Known media types with fallback for vendor-specific MIME values. */
sealed class CCDAMediaType(open val mimeType: String?) {
    /** PNG image. */
    data object ImagePNG : CCDAMediaType("image/png")
    /** JPEG image. */
    data object ImageJPEG : CCDAMediaType("image/jpeg")
    /** GIF image. */
    data object ImageGIF : CCDAMediaType("image/gif")
    /** PDF document. */
    data object ApplicationPDF : CCDAMediaType("application/pdf")
    /** Plain text document. */
    data object TextPlain : CCDAMediaType("text/plain")
    /** HTML document. */
    data object TextHTML : CCDAMediaType("text/html")
    /** Unsupported MIME value preserved from source XML. */
    data class Unsupported(override val mimeType: String) : CCDAMediaType(mimeType)
    /** Missing or empty MIME value. */
    data object Unknown : CCDAMediaType(null)

    companion object {
        /** Creates a media type from a MIME string. */
        fun fromMimeType(mimeType: String?): CCDAMediaType = when (mimeType?.lowercase()) {
            null, "" -> Unknown
            "image/png" -> ImagePNG
            "image/jpeg", "image/jpg" -> ImageJPEG
            "image/gif" -> ImageGIF
            "application/pdf" -> ApplicationPDF
            "text/plain" -> TextPlain
            "text/html" -> TextHTML
            else -> Unsupported(mimeType)
        }
    }
}
